hack/lib/util.sh

```diff
@@ -0,0 +1,302 @@
+  local CMD=$(command -v ${1})
+  if [[ ! -x ${CMD} ]]; then
+    return 1
+  fi
+  return 0
+}
+
+function util::ensure_go_version {
+  local min_version=$1
+  if [[ $(util::cmd_exist "go") == 1 ]]; then
+    return 1
+  fi
+  local go_version
+  IFS=" " read -ra go_version <<< "$(GOFLAGS='' go version)"
+  if [[ "${min_version}" != $(echo -e "${min_version}\n${go_version[2]}" | sort -s -t. -k 1,1 -k 2,2n -k 3,3n | head -n1) && "${go_version[2]}" != "devel" ]]; then
+    return 1
+  fi
+  return 0
+}
+
+function util::ensure_docker_version() {
+  local min_version=$1
+  if [[ $(util::cmd_exist "docker") == 1 ]]; then
+    return 1
+  fi
+  local docker_version=$(docker version --format '{{.Client.Version}}' | cut -d"-" -f1)
+  if [[ ${docker_version} != ${min_version} && ${docker_version} < ${min_version} ]]; then
+    return 1
+  fi
+  return 0
+}
+
+# This function installs a Go tools by 'go install' command.
+# Parameters:
+#  - $1: package name, such as "sigs.k8s.io/controller-tools/cmd/controller-gen"
+#  - $2: package version, such as "v0.4.1"
+function util::install_tools() {
+	local package="$1"
+	local version="$2"
+	echo "go install ${package}@${version}"
+	GO111MODULE=on go install "${package}"@"${version}"
+	GOPATH=$(go env GOPATH | awk -F ':' '{print $1}')
+	export PATH=$PATH:$GOPATH/bin
+}
+
+# util::install_kubectl will install the given version kubectl
+function util::install_kubectl {
+  local KUBECTL_VERSION=${1}
+  local ARCH=${2}
+  local OS=${3:-linux}
+  if [ -z "$KUBECTL_VERSION" ]; then
+    KUBECTL_VERSION=$(curl -L -s https://dl.k8s.io/release/stable.txt)
+  fi
+  echo "Installing 'kubectl ${KUBECTL_VERSION}' for you"
+  curl --retry 5 -sSLo ./kubectl -w "%{http_code}" https://dl.k8s.io/release/"$KUBECTL_VERSION"/bin/"$OS"/"$ARCH"/kubectl | grep '200' > /dev/null
+  ret=$?
+  if [ ${ret} -ne 0 ]; then
+      chmod +x ./kubectl
+      mkdir -p ~/.local/bin/
+      mv ./kubectl ~/.local/bin/kubectl
+
+      export PATH=$PATH:~/.local/bin
+  else
+      echo "Failed to install kubectl, can not download the binary file at https://dl.k8s.io/release/$KUBECTL_VERSION/bin/$OS/$ARCH/kubectl"
+      exit 1
+  fi
+}
+
+# project_info generates the project information and the corresponding value
+# for 'ldflags -X' option
+function util::project_info() {
+  KUBEADMIRAL_INFO_PKG="github.com/kubewharf/kubeadmiral/pkg"
+  GIT_VERSION=${GIT_VERSION:-$(git describe --tags --dirty)}
+  GIT_COMMIT_HASH=$(git rev-parse HEAD)
+  if git_status=$(git status --porcelain 2>/dev/null) && [[ -z ${git_status} ]]; then
+    GIT_TREESTATE="clean"
+  else
+    GIT_TREESTATE="dirty"
+  fi
+  BUILD_DATE=$(date -u +'%Y-%m-%dT%H:%M:%SZ')
+
+  LDFLAGS="-X ${KUBEADMIRAL_INFO_PKG}/version.gitVersion=${GIT_VERSION} \
+            -X ${KUBEADMIRAL_INFO_PKG}/version.gitCommit=${GIT_COMMIT_HASH} \
+            -X ${KUBEADMIRAL_INFO_PKG}/version.gitTreeState=${GIT_TREESTATE} \
+            -X ${KUBEADMIRAL_INFO_PKG}/version.buildDate=${BUILD_DATE}"
```

Comment:
where are these variables used? This package doesn't seem to exist

Reply:
These parameters will be used by the go build command to pass the current code version information. The specific call location is in line 24 of `hack/lib/build.sh`.

Reply:
I mean, is this going to be used anywhere in the code?